modules/tinymce/src/plugins/anchor/main/ts/core/Anchor.ts

```diff
@@ -5,33 +5,74 @@
  * For commercial licenses see https://www.tiny.cloud/
  */
 
+import { HTMLAnchorElement } from '@ephox/dom-globals';
 import Editor from 'tinymce/core/api/Editor';
-import { Element } from '@ephox/dom-globals';
+import RangeUtils from 'tinymce/core/api/dom/RangeUtils';
+import Tools from 'tinymce/core/api/util/Tools';
+import * as Settings from '../api/Settings';
+import * as Utils from './Utils';
 
-const isNamedAnchor = (editor: Editor, node: Element) => node.tagName === 'A' && editor.dom.getAttrib(node, 'href') === '';
+const removeEmptyNamedAnchorsInSelection = (editor: Editor) => {
+  const dom = editor.dom;
+  RangeUtils(dom).walk(editor.selection.getRng(), (nodes) => {
+    Tools.each(nodes, (node) => {
+      if (Utils.isEmptyNamedAnchor(node)) {
+        dom.remove(node, false);
+      }
+    });
+  });
+};
 
 const isValidId = (id: string) =>
   // Follows HTML4 rules: https://www.w3.org/TR/html401/types.html#type-id
   /^[A-Za-z][A-Za-z0-9\-:._]*$/.test(id);
 
+const getNamedAnchor = (editor: Editor): HTMLAnchorElement | null =>
+  editor.dom.getParent(editor.selection.getStart(), Utils.namedAnchorSelector) as HTMLAnchorElement;
+
 const getId = (editor: Editor) => {
-  const selectedNode = editor.selection.getNode();
-  return isNamedAnchor(editor, selectedNode) ? (selectedNode.getAttribute('id') || selectedNode.getAttribute('name')) : '';
+  const anchor = getNamedAnchor(editor);
+  if (anchor) {
+    return Utils.getIdFromAnchor(anchor);
+  } else {
+    return '';
+  }
 };
 
-const insert = (editor: Editor, id: string) => {
-  const selectedNode = editor.selection.getNode();
+const createAnchor = (editor: Editor, id: string) => {
+  editor.undoManager.transact(() => {
+    if (!Settings.allowHtmlInNamedAnchor(editor)) {
+      editor.selection.collapse(true);
+    }
+    if (editor.selection.isCollapsed()) {
+      editor.insertContent(editor.dom.createHTML('a', { id }));
+    } else {
+      // Remove any empty named anchors in the selection as they cannot be removed by the formatter since they are cef
+      removeEmptyNamedAnchorsInSelection(editor);
+      // Format is set up to truncate any partially selected named anchors so that they are not completely removed
+      editor.formatter.remove('namedAnchor', null, null, true);
+      // Insert new anchor using the formmater - will wrap selected content in anchor
+      editor.formatter.apply('namedAnchor', { value: id });
+      // Need to add visual classes to anchors if required
+      editor.addVisual();
+    }
+  });
+};
+
+const updateAnchor = (editor: Editor, id: string, anchorElement: HTMLAnchorElement): void => {
+  anchorElement.removeAttribute('name');
+  anchorElement.id = id;
+  editor.addVisual(); // Need to add visual classes to anchors if required
```

Comment:
Shouldn't this already have a visual marker? I guess it doesn't hurt to make sure :shrug:

Reply:
In most cases it should. Although, I did find one situation where it didn't when converting a bare anchor to a named anchor. I  figured it also ensures all anchors are displaying correctly.